src/runtime/capi.cpp

```diff
@@ -1034,14 +1053,145 @@ extern "C" PyObject* PyErr_Occurred() noexcept {
+        const char* file_str = PyString_AsString(*filename);
+        if (file_str == NULL || (len < 0 && PyErr_Occurred()))
+            goto handle_error;
+
+        /* if filename.lower().endswith((".pyc", ".pyo")): */
+        if (len >= 4 && file_str[len - 4] == '.' && tolower(file_str[len - 3]) == 'p'
+            && tolower(file_str[len - 2]) == 'y'
+            && (tolower(file_str[len - 1]) == 'c' || tolower(file_str[len - 1]) == 'o')) {
+            *filename = PyString_FromStringAndSize(file_str, len - 1);
+            if (*filename == NULL)
+                goto handle_error;
+        } else
+            Py_INCREF(*filename);
+    } else {
+        const char* module_str = PyString_AsString(*module);
+        *filename = NULL;
+        if (module_str && strcmp(module_str, "__main__") == 0) {
+            PyObject* argv = PySys_GetObject("argv");
+            if (argv != NULL && PyList_Size(argv) > 0) {
+                int is_true;
+                *filename = PyList_GetItem(argv, 0);
+                Py_INCREF(*filename);
+                /* If sys.argv[0] is false, then use '__main__'. */
+                is_true = PyObject_IsTrue(*filename);
+                if (is_true < 0) {
+                    Py_DECREF(*filename);
+                    goto handle_error;
+                } else if (!is_true) {
+                    Py_DECREF(*filename);
+                    *filename = PyString_FromString("__main__");
+                    if (*filename == NULL)
+                        goto handle_error;
+                }
+            } else {
+                /* embedded interpreters don't have sys.argv, see bug #839151 */
+                *filename = PyString_FromString("__main__");
+                if (*filename == NULL)
+                    goto handle_error;
+            }
+        }
+        if (*filename == NULL) {
+            *filename = *module;
+            Py_INCREF(*filename);
+        }
+    }
+
+    return 1;
+
+handle_error:
+    /* filename not XDECREF'ed here as there is no way to jump here with a
+       dangling reference. */
+    Py_XDECREF(*registry);
+    Py_XDECREF(*module);
+    return 0;
+}
+
+static PyObject* do_warn(PyObject* message, PyObject* category, Py_ssize_t stack_level) {
+    PyObject* filename, *module, *registry, *res;
+    int lineno;
+
+    if (!setup_context(stack_level, &filename, &lineno, &module, &registry))
+        return NULL;
+
+    res = warn_explicit(category, message, filename, lineno, module, registry, NULL);
+    Py_DECREF(filename);
+    Py_DECREF(registry);
+    Py_DECREF(module);
+    return res;
+}
+
 extern "C" int PyErr_WarnEx(PyObject* category, const char* text, Py_ssize_t stacklevel) noexcept {
     // These warnings are silenced by default:
     // We should copy the real CPython code in here
     if (category == PyExc_DeprecationWarning)
         return 0;
 
-    fatalOrError(PyExc_NotImplementedError, "unimplemented");
-    return -1;
+    PyObject* res;
```

Comment:
Is it possible to get rid of the first part of this function now?  (It's a workaround that I added, not something from CPython.)